evm/src/cpu/kernel/asm/bignum.asm

```diff
@@ -0,0 +1,296 @@
+    // stack: a_len-i-1 == 0, b_len-i-1 == 0, a_len-i-1, b_len-i-1, a_i_loc+1, b_i_loc+1, retdest
+    %stack (all: 6) -> (1)
+    // stack: 1, retdest
+    SWAP1
+    JUMP
+
+// Replaces a with a + b, where a and b are unbounded big-endian integers represented with one-byte limbs.
+// Leave b unchanged.
+global add_bignum:
+    // stack: a_len, b_len, a_start_loc, b_start_loc, retdest
+    %stack (al, bl, a, b) -> (al, a, bl, b, bl)
+    // stack: a_len, a_start_loc, b_len, b_start_loc, b_len, retdest
+    ADD
+    %decrement
+    // stack: a_end_loc, b_len, b_start_loc, b_len, retdest
+    %stack (a, bl, b, bl) -> (bl, b, bl, a)
+    // stack: b_len, b_start_loc, b_len, a_end_loc, retdest
+    ADD
+    %decrement
+    // stack: b_end_loc, b_len, a_end_loc, retdest
+    %stack (be, bl, ae) -> (0, 0, ae, be, bl)
+    // stack: carry=0, i=0, a_end_loc, b_end_loc, n=b_len, retdest
+add_loop:
+    // stack: carry, i, a_i_loc, b_i_loc, n, retdest
+    DUP4
+    DUP4
+    // stack: a_i_loc, b_i_loc, carry, i, a_i_loc, b_i_loc, n, retdest
+    %mload_kernel_general
+    SWAP1
+    %mload_kernel_general
+    SWAP1
+    // stack: a[i], b[i], carry, i, a_i_loc, b_i_loc, n, retdest
+    ADD
+    ADD
+    // stack: a[i] + b[i] + carry, i, a_i_loc, b_i_loc, n, retdest
+    %stack (val) -> (val, 256, 256, val)
+    // stack: a[i] + b[i] + carry, 256, 256, a[i] + b[i] + carry, i, a_i_loc, b_i_loc, n, retdest
+    DIV
+    // stack: (a[i] + b[i] + carry) // 256, 256, a[i] + b[i] + carry, i, a_i_loc, b_i_loc, n, retdest
+    SWAP2
+    // stack: a[i] + b[i] + carry, 256, (a[i] + b[i] + carry) // 256, i, a_i_loc, b_i_loc, n, retdest
+    MOD
+    // stack: c[i] = (a[i] + b[i] + carry) % 256, carry_new = (a[i] + b[i] + carry) // 256, i, a_i_loc, b_i_loc, n, retdest
+    DUP4
+    // stack: a_i_loc, c[i], carry_new, i, a_i_loc, b_i_loc, n, retdest
+    %mstore_kernel_general
+    // stack: carry_new, i, a_i_loc, b_i_loc, n, retdest
+    %stack (c, i, a, b) -> (a, b, c, i)
+    // stack: a_i_loc, b_i_loc, carry_new, i, n, retdest
+    %decrement
+    SWAP1
+    %decrement
+    SWAP1
+    %stack (a, b, c, i) -> (c, i, a, b)
+    // stack: carry_new, i, a_i_loc - 1, b_i_loc - 1, n, retdest
+    SWAP1
+    %increment
+    SWAP1
+    // stack: carry_new, i + 1, a_i_loc - 1, b_i_loc - 1, n, retdest
+    DUP5
+    DUP3
+    // stack: i + 1, n, carry_new, i + 1, a_i_loc - 1, b_i_loc - 1, n, retdest
+    EQ
+    %not_bool
+    %jumpi(add_loop)
+add_end:
+    STOP
+    // stack: carry_new, i + 1, a_i_loc - 1, b_i_loc - 1, n, retdest
+    %stack (c, i, a, b, n) -> (c, a)
+    // stack: carry_new, a_i_loc - 1, retdest
+    // If carry = 0, no need to decrement.
+    ISZERO
+    %jumpi(increment_end)
+increment_loop:
+    // stack: cur_loc, retdest
+    DUP1
+    %mload_kernel_general
+    // stack: val, cur_loc, retdest
+    %increment
+    // stack: val+1, cur_loc, retdest
+    %stack (v, l) -> (l, v, l, v)
+    DUP2
+    // stack: cur_loc, val+1, cur_loc, val+1, retdest
```

Comment:
This doesn’t make sense. If the stack is `cur_loc, val+1, cur_loc, val+1` and then you do a `DUP2`, then the new stack is `val+1, cur_loc, val+1, cur_loc, val+1`.

Reply:
This edge case should have a unit test.